File: scripts/launch.ts
#!/usr/bin/env bun
// One command instead of the four-terminal-plus-manual-paste dance in
// CLAUDE.md's "Running things for manual/browser testing": starts
// hexend, the editor, and presentation, waits for each to come up, and
// opens a browser tab for the editor and one for presentation, both
// already pointed at the given project via query params — the
// editor's own "?path=" (and "?server="), see App.tsx/server.ts, and
// presentation's pre-existing "?server=&path=", see PresentationApp.tsx.
//
// Usage: bun run launch --project examples/demo/demo.hexen.yml
// Must run inside `nix develop` — same as every other hexend-touching
// command in this repo (protoc/buf/cargo aren't on PATH otherwise).
import { resolve } from "node:path";

function parseArgs(argv: string[]) {
  const args: Record<string, string> = {};
  const flags = new Set<string>();
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]!;
    if (!arg.startsWith("--")) throw new Error(`Expected a --flag, got "${arg}"`);
    const name = arg.slice(2);
    const value = argv[i + 1];
    if (value === undefined || value.startsWith("--")) {
      flags.add(name);
      continue;
    }
    args[name] = value;
    i++;
  }
  return { args, flags };
}

function usageError(message: string): never {
  console.error(message);
  console.error(
    "Usage: bun run launch --project <file.hexen.yml> [--hexend-port 4000] [--editor-port 5173] [--presentation-port 5174] [--no-browser]",
  );
  process.exit(1);
}

interface Service {
  name: string;
  proc: ReturnType<typeof Bun.spawn>;
}

const services: Service[] = [];
let shuttingDown = false;

function spawnService(name: string, cmd: string[], opts: { cwd?: string; env?: Record<string, string> } = {}): ReturnType<typeof Bun.spawn> {
  console.log(`$ ${cmd.join(" ")}${opts.cwd ? ` (in ${opts.cwd})` : ""}`);
  const proc = Bun.spawn(cmd, {
    cwd: opts.cwd,
    env: { ...process.env, ...opts.env },
    stdout: "inherit",
    stderr: "inherit",
  });
  services.push({ name, proc });
  return proc;
}

/** Polls until something answers on `url`, or throws after `timeoutMs`. Doesn't care about the response's status — a listening socket is all "up" means here. */
async function waitForHttp(name: string, url: string, timeoutMs: number): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    try {
      await fetch(url);
      return;
    } catch {
      await new Promise((r) => setTimeout(r, 200));
    }
  }
  throw new Error(`Timed out waiting for ${name} at ${url}`);
}

function openBrowser(url: string): void {
  const cmd =
    process.platform === "darwin"
      ? ["open", url]
      : process.platform === "win32"
        ? ["cmd", "/c", "start", "", url]
        : ["xdg-open", url];
  try {
    Bun.spawn(cmd, { stdout: "ignore", stderr: "ignore" }).unref();
  } catch {
    console.warn(`Couldn't open a browser automatically — open this yourself:\n  ${url}`);
  }
}

function killAll(): void {
  if (shuttingDown) return;
  shuttingDown = true;
  for (const { proc } of services) {
    try {
      proc.kill();
    } catch {
      // already gone
    }
  }
}

async function main() {
  const { args, flags } = parseArgs(process.argv.slice(2));

  const projectArg = args["project"];
  if (!projectArg) usageError("Missing --project <file.hexen.yml>");
  const projectPath = resolve(projectArg);
  if (!(await Bun.file(projectPath).exists())) usageError(`No such file: ${projectPath}`);

  const hexendPort = Number(args["hexend-port"] ?? 4000);
  const editorPort = Number(args["editor-port"] ?? 5173);
  const presentationPort = Number(args["presentation-port"] ?? 5174);
  const openBrowsers = !flags.has("no-browser");

  process.on("SIGINT", () => {
    killAll();
    process.exit(0);
  });
  process.on("SIGTERM", () => {
    killAll();
    process.exit(0);
  });

  console.log(`Starting hexend on :${hexendPort}...`);
  spawnService("hexend", ["cargo", "run", "--manifest-path", "apps/hexend/Cargo.toml"], {
    env: { PORT: String(hexendPort) },
  });
  await waitForHttp("hexend", `http://localhost:${hexendPort}/`, 60_000);

  console.log(`Starting the editor on :${editorPort}...`);
  spawnService("editor", ["bun", "run", "--cwd", "apps/editor", "dev", "--", "--port", String(editorPort), "--strictPort"]);
  await waitForHttp("editor", `http://localhost:${editorPort}/`, 30_000);

  console.log(`Starting presentation on :${presentationPort}...`);
  spawnService("presentation", [
    "bun",
    "run",
    "--cwd",
    "apps/presentation",
    "dev",
    "--",
    "--port",
    String(presentationPort),
    "--strictPort",
  ]);
  await waitForHttp("presentation", `http://localhost:${presentationPort}/`, 30_000);

  const server = `http://localhost:${hexendPort}`;
  const editorUrl = `http://localhost:${editorPort}/?server=${encodeURIComponent(server)}&path=${encodeURIComponent(projectPath)}`;
  const presentationUrl = `http://localhost:${presentationPort}/?server=${encodeURIComponent(server)}&path=${encodeURIComponent(projectPath)}`;

  console.log("\nEverything's up:");
  console.log(`  editor:       ${editorUrl}`);
  console.log(`  presentation: ${presentationUrl}`);

  if (openBrowsers) {
    openBrowser(editorUrl);
    openBrowser(presentationUrl);
  }

  console.log("\nPress Ctrl+C to stop everything.");
  await Promise.race(services.map(({ proc }) => proc.exited));
  // One of them died on its own (a crash, or a port fight) — bring the rest down too rather than leaving orphans.
  killAll();
}

main().catch((err) => {
  console.error(err instanceof Error ? err.message : String(err));
  killAll();
  process.exit(1);
});

File: apps/hexen-cli/src/launch.rs
//! `hexen launch` — one command for the local dev loop (port of master's
//! scripts/launch.ts): starts hexend and trunk for the editor and the
//! presentation view, waits for each to come up, then opens a browser
//! tab for each already pointed at the project. When any one of them
//! exits, the rest are stopped too. Ctrl+C reaches them all directly
//! (they share the terminal's process group).

use std::net::{SocketAddr, TcpStream};
use std::path::{Path, PathBuf};
use std::process::{Child, Command};
use std::thread::sleep;
use std::time::{Duration, Instant};

pub struct Ports {
    pub hexend: u16,
    pub editor: u16,
    pub presentation: u16,
}

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../..")
}

/// Everything launch started, stopped when this is dropped.
struct Services(Vec<(&'static str, Child)>);

impl Drop for Services {
    fn drop(&mut self) {
        for (_, child) in &mut self.0 {
            let _ = child.kill();
            let _ = child.wait();
        }
    }
}

impl Services {
    fn spawn(&mut self, name: &'static str, command: &mut Command) -> Result<(), String> {
        let child = command
            .spawn()
            .map_err(|e| format!("couldn't start {name}: {e}"))?;
        self.0.push((name, child));
        Ok(())
    }

    /// Waits until something accepts connections on `port`, or fails if
    /// the service exits first or `timeout` passes. (A first trunk build
    /// compiles the whole app to wasm, so its timeout is long.)
    fn wait_for(&mut self, name: &str, port: u16, timeout: Duration) -> Result<(), String> {
        let addr = SocketAddr::from(([127, 0, 0, 1], port));
        let deadline = Instant::now() + timeout;
        while Instant::now() < deadline {
            if TcpStream::connect_timeout(&addr, Duration::from_millis(200)).is_ok() {
                return Ok(());
            }
            if let Some(exited) = self.first_exited() {
                return Err(format!("{exited} exited before {name} came up"));
            }
            sleep(Duration::from_millis(300));
        }
        Err(format!("timed out waiting for {name} on port {port}"))
    }

    fn first_exited(&mut self) -> Option<&'static str> {
        self.0
            .iter_mut()
            .find_map(|(name, child)| matches!(child.try_wait(), Ok(Some(_))).then_some(*name))
    }
}

fn open_browser(url: &str) {
    let mut command = if cfg!(target_os = "macos") {
        Command::new("open")
    } else if cfg!(windows) {
        let mut c = Command::new("cmd");
        c.args(["/c", "start", ""]);
        c
    } else {
        Command::new("xdg-open")
    };
    if command.arg(url).spawn().is_err() {
        eprintln!("Couldn't open a browser automatically — open this yourself:\n  {url}");
    }
}

fn encode(s: &str) -> String {
    let mut out = String::new();
    for b in s.bytes() {
        if b.is_ascii_alphanumeric() || b"-_.~".contains(&b) {
            out.push(b as char);
        } else {
            out.push_str(&format!("%{b:02X}"));
        }
    }
    out
}

/// The editor and presentation URLs for a project on a hexend.
pub fn app_urls(project: &Path, ports: &Ports) -> (String, String) {
    let server = format!("http://localhost:{}", ports.hexend);
    let query = format!(
        "?server={}&path={}",
        encode(&server),
        encode(&project.to_string_lossy())
    );
    (
        format!("http://localhost:{}/{query}", ports.editor),
        format!("http://localhost:{}/{query}", ports.presentation),
    )
}

pub fn run(project: &Path, ports: Ports, open: bool) -> Result<(), String> {
    let project = std::path::absolute(project).map_err(|e| e.to_string())?;
    if !project.is_file() {
        return Err(format!("No such file: {}", project.display()));
    }
    let root = repo_root();
    let mut services = Services(Vec::new());

    println!("Starting hexend on :{}…", ports.hexend);
    services.spawn(
        "hexend",
        Command::new("cargo")
            .args(["run", "-p", "hexend"])
            .current_dir(&root)
            .env("PORT", ports.hexend.to_string()),
    )?;
    services.wait_for("hexend", ports.hexend, Duration::from_secs(300))?;

    for (name, dir, port) in [
        ("the editor", "apps/editor", ports.editor),
        ("presentation", "apps/presentation", ports.presentation),
    ] {
        println!("Starting {name} on :{port}…");
        services.spawn(
            if dir == "apps/editor" {
                "editor"
            } else {
                "presentation"
            },
            Command::new("trunk")
                .args(["serve", "--port", &port.to_string()])
                .current_dir(root.join(dir)),
        )?;
        services.wait_for(name, port, Duration::from_secs(600))?;
    }

    let (editor, presentation) = app_urls(&project, &ports);
    println!("\nEverything's up:\n  editor:       {editor}\n  presentation: {presentation}");
    if open {
        open_browser(&editor);
        open_browser(&presentation);
    }

    println!("\nPress Ctrl+C to stop everything.");
    loop {
        if let Some(name) = services.first_exited() {
            // A crash or a port fight — bring the rest down rather than leave orphans.
            return Err(format!("{name} exited; stopping the rest"));
        }
        sleep(Duration::from_millis(500));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn urls_carry_the_server_and_the_absolute_project_path() {
        let ports = Ports {
            hexend: 4001,
            editor: 5173,
            presentation: 5174,
        };
        let (editor, presentation) = app_urls(Path::new("/maps/my realm.hexen.yml"), &ports);
        assert_eq!(
            editor,
            "http://localhost:5173/?server=http%3A%2F%2Flocalhost%3A4001&path=%2Fmaps%2Fmy%20realm.hexen.yml"
        );
        assert!(presentation.starts_with("http://localhost:5174/?server="));
    }
}
